File: src/test/java/com/snooper/service/CommandHandlerTest.java
/*
 * Copyright 2015-2018 the original author or authors.
 *
 * All rights reserved. This program and the accompanying materials are
 * made available under the terms of the Eclipse Public License v2.0 which
 * accompanies this distribution and is available at
 *
 * http://www.eclipse.org/legal/epl-v20.html
 */

package com.snooper.service;

import com.snooper.constant.Action;
import com.snooper.constant.Direction;
import com.snooper.dto.Plateau;
import com.snooper.dto.RoboticRover;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class CommandHandlerTest {

	private CommandHandlerImpl commandHandler = new CommandHandlerImpl();

	@Test
	void parsePlateauShouldReturnPlateau() {
		Plateau plateau = commandHandler.parsePlateau("4 4");

		assertEquals(4, plateau.getMaxX());
		assertEquals(4, plateau.getMaxY());
		assertEquals(0, plateau.getMinY());
		assertEquals(0, plateau.getMinY());
	}

	@Test
	void parsePlateauShouldThrowError() {
		Exception exception = assertThrows(IllegalArgumentException.class, () ->
				commandHandler.parsePlateau("4"));

		assertEquals("Invalid command for plateau coordinates", exception.getMessage());
	}

	@Test
	void parseRoboticRoverShouldReturnRoboticRover() {
		RoboticRover roboticRover = commandHandler.parseRoboticRover("4 4 N");

		assertEquals(4, roboticRover.getCoordinate().getX());
		assertEquals(4, roboticRover.getCoordinate().getY());
		assertEquals(Direction.valueOf("N"), roboticRover.getDirection());
	}

	@Test
	void parseRoboticRoverShouldThrowError() {
		Exception exception = assertThrows(IllegalArgumentException.class, () ->
				commandHandler.parseRoboticRover("4 4")
		);

		assertTrue(exception.getMessage().contains("roboticRover"));
	}

	@Test
	void parseActionsShouldReturnActions() {
		List<Action> actions = commandHandler.parseActions("LRM");

		List<Action> expectations = Arrays.asList(Action.L, Action.R, Action.M);
		for (int i = 0; i < actions.size(); i++) {
			assertEquals(expectations.get(i), actions.get(i));
		}
	}

	@Test
	void parseActionsShouldThrowError() {
		Exception exception = assertThrows(IllegalArgumentException.class, () ->
				commandHandler.parseActions(null)
		);
		assertTrue(exception.getMessage().contains("actions"));
	}
}

File: src/test/java/com/snooper/service/CommandHandlerTest.kt
package com.snooper.service

import com.snooper.constant.Action
import com.snooper.constant.Direction
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

internal class CommandHandlerTest {
    private val commandHandler = CommandHandlerImpl()

    @Test
    fun `parsePlateau should return plateau`() {
        val (_, minY, maxX, maxY) = commandHandler.parsePlateau("4 4")
        assertEquals(4, maxX)
        assertEquals(4, maxY)
        assertEquals(0, minY)
        assertEquals(0, minY)
    }

    @Test
    fun `parsePlateau should throw exception when command is incorrect`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java) {
            commandHandler.parsePlateau("4") }

        assertEquals("Invalid command for plateau coordinates", exception.message)
    }

    @Test
    fun `parsePlateau should throw exception when wrong coordinates format`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java) {
            commandHandler.parsePlateau("S 4") }

        assertEquals("Invalid plateau coordinates", exception.message)
    }

    @Test
    fun `parsePlateau should throw exception when wrong minX and maxX`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java) {
            commandHandler.parsePlateau("-1 1") }

        assertEquals("Invalid minX and maxX", exception.message)
    }

    @Test
    fun `parsePlateau should throw exception when wrong minY and maxY`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java) {
            commandHandler.parsePlateau("1 -1") }

        assertEquals("Invalid minY and maxY", exception.message)
    }

    @Test
    fun `parseRoboticRover should return roboticRover`() {
        val (coordinate, direction) = commandHandler.parseRoboticRover("4 4 N")

        assertEquals(4, coordinate.x)
        assertEquals(4, coordinate.y)
        assertEquals(Direction.valueOf("N"), direction)
    }

    @Test
    fun `parseRoboticRover should throw exception when command is missing direction`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java
        ) { commandHandler.parseRoboticRover("4 4") }

        assertTrue(exception.message!!.contains("roboticRover"))
    }

    @Test
    fun `parseRoboticRover should throw exception when command is using wrong direction`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java
        ) { commandHandler.parseRoboticRover("4 4 T") }

        assertTrue(exception.message!!.contains("Invalid direction"))
    }

    @Test
    fun `parseRoboticRover should throw exception when command is using wrong x or y`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java
        ) { commandHandler.parseRoboticRover("S 4 T") }

        assertTrue(exception.message!!.contains("Invalid roboticRover coordinates"))
    }

    @Test
    fun `parseActions should return actions`() {
        val actions = commandHandler.parseActions("LRM")
        val expectations = listOf(Action.L, Action.R, Action.M)

        for (i in actions.indices) {
            assertEquals(expectations[i], actions[i])
        }
    }

    @Test
    fun `parseActions should throw error when command is using wrong action`() {
        val exception: Exception = assertThrows(IllegalArgumentException::class.java) {
            commandHandler.parseActions("T") }

        assertTrue(exception.message!!.contains("Invalid action"))
    }
}